libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java

```diff
@@ -328,6 +340,62 @@ private static PolicyManager createPolicyManager() {
         );
     }
 
+    private static Set<Path> pathSet(Path... paths) {
+        return Arrays.stream(paths).map(x -> x.toAbsolutePath().normalize()).collect(Collectors.toUnmodifiableSet());
+    }
+
+    // package visible for tests
+    static void validateFilesEntitlements(
+        Map<String, Policy> pluginPolicies,
+        PathLookup pathLookup,
+        Path configDir,
+        Path pluginsDir,
+        Path modulesDir,
+        Path libDir
```

Comment:
Optional: I'd consider simply passing in the path sets here, rather than all individual dirs.

Soon we'll be changing `bootstrapArgs` so it doesn't have N individual dir fields, but rather has a single getter that accepts an enum parameter. Making `validateFilesEntitlements` accept sets would probably work more smoothly once this change happens.

As an example: consider that during unit testing, there is usually no `configDir`.

I do see the benefit of having this method have the responsibility for determining which dirs are read/write forbidden, though, so I understand the inclination to pass them all in the way you did.

Reply:
Actually, based on my additional comments below... what about passing a pair of `FileAccessTree`s?

Reply:
I used a FileAccessTree in the end, for the corner cases, but that does not help with the paths to check.